From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zach Brown <zach.brown@destroystokyo.com>
Date: Tue, 4 Sep 2018 19:07:57 -0400
Subject: [PATCH] Origami: MC-2025: Save and load entity AABB to prevent wobble

What follows is a summarized analysis provided on the Mojira and associated subreddit by various MC community members
who have investigated this issue and found a solution. This work is largely a result of their efforts.

The underlying cause of MC-2025 is that sometimes an AABB ends up slightly smaller than the desired width. If this
happens before the entity is pushed up against a boundary (i.e. blocks or walls), then upon chunk save and reload, the
AABB size will be recomputed such that it is intersecting the wall, allowing the entity to be pushed into the wall,
suffocate, and die. Although the rounding artifacts get larger at larger world coordinates, the drift we see is
miniscule. Closer to the world origin, we have seen error on the order of 2-46. Compare this to the fact that
(due to MC-4), entity coordinates send to clients are quantized to multiples of 1/4096 (2-12).

But, OMG, do the rounding errors mean that AABB's accumulate shrinkage over time? Actually, no. The statistics on IEEE
rounding do not have that kind of bias. What has not been stated is that the AABB is just as likely to end up larger
than the expected width; on save and reload, the entity ends up slightly away from the wall, and we don't notice any
problem. In reality, the AABB size ends up undergoing random wobble around the expected value all the time, and that
wobble isn't functionally any different from the kind we'd get even if we tried to force the AABB size to be stable!

This reasoning leads us to one clear conclusion: The simplest, least invasive, and most correct solution is to just
store the AABB in NBT data on chunk save and restore the AABB exactly as it was saved upon reload.

Original code by Zach Brown, licensed under GNU Lesser General Public License v3.0
you can find the original code on https://github.com/Minebench/Origami/

Updated to 1.17 by: notsatvrn <satvrn@gmx.com>

diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 3fd3eb4d20c6e4c65b4c0924fad529da2311204f..bbdd9a004c7b233ce9517a78808034047dcaec3e 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2184,6 +2184,12 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
             if (fromNetherPortal) {
                 nbt.putBoolean("Paper.FromNetherPortal", true);
             }
+            // Paper start - MC-2025 fix - Save entity AABB and load it, floating point issues recalculating AABB can result in wobble
+            AABB boundingBox = this.getBoundingBox();
+            nbt.put("Paper.AAAB", this.newDoubleList(
+                boundingBox.minX, boundingBox.minY, boundingBox.minZ,
+                boundingBox.maxX, boundingBox.maxY, boundingBox.maxZ
+            ));
             // Paper end
             // Origami start
             if (fromMobFarm) {
@@ -2271,6 +2277,16 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
             } else {
                 throw new IllegalStateException("Entity has invalid position");
             }
+            // Paper start - MC-2025 fix - Save entity AABB and load it, floating point issues recalculating AABB can result in wobble
+            // Placement is important, always after the setPosition call above
+            if (nbt.contains("Paper.AABB")) {
+                ListTag savedBB = nbt.getList("Paper.AABB", 6);
+                this.setBoundingBox(new AABB(
+                    savedBB.getDouble(0), savedBB.getDouble(1), savedBB.getDouble(2),
+                    savedBB.getDouble(3), savedBB.getDouble(4), savedBB.getDouble(5)
+                ));
+            }
+            // Paper end
 
             // CraftBukkit start
             if (this instanceof net.minecraft.world.entity.LivingEntity) {
